src/python/Utils/CPMetrics.py

```diff
@@ -0,0 +1,305 @@
+# TYPE cherrypy_http_server_requests counter
+cherrypy_http_server_requests -1
+# HELP cherrypy_app_total_bytes_written
+# TYPE cherrypy_app_total_bytes_written counter
+cherrypy_app_total_bytes_written 0
+# HELP cherrypy_http_server_read_throughput
+# TYPE cherrypy_http_server_read_throughput counter
+cherrypy_http_server_read_throughput -1
+# HELP cherrypy_http_server_accepts
+# TYPE cherrypy_http_server_accepts counter
+cherrypy_http_server_accepts 0
+"""
+# python3 compatibility
+from __future__ import division
+
+# system modules
+import os
+import json
+
+import cherrypy
+from cherrypy.lib import cpstats
+
+
+def promMetrics(exporter_name):
+    """
+    Provide cherrypy stats prometheus metrics for given exporter name.
+    """
+    data = cpstats.StatsPage().data()
+    metrics = flattenStats(data)
+    if isinstance(metrics, str):
+        metrics = json.loads(metrics)
+    # the following keys will be skipped
+    skip = [
+        'cherrypy_app_enabled',
+        'cherrypy_http_server_bind_address',
+        'cherrypy_app_requests',
+        'cherrypy_app_server_version',
+        'cherrypy_http_server_enabled'
+    ]
+    # our prometheus data representation
+    pdata = ""
+    for key, val in metrics.iteritems():
+        if key in skip:
+            continue
+        # add exporter name as a prefix for each key
+        key = '{}_{}'.format(exporter_name, key)
+        mhelp = "# HELP {}\n".format(key)
+        if isinstance(val, list):
+            mtype = "# TYPE {} histogram\n".format(key)
+            pdata += mhelp
+            pdata += mtype
+            for wdict in val:
+                entries = []
+                for kkk, vvv in wdict.iteritems():
+                    entries.append('{}="{}"'.format(kkk, vvv))
+                entry = "{%s}" % ','.join(entries)
+                pdata += "{}{} 1\n".format(key, entry)
+        elif isinstance(val, (str, tuple)):
+            continue
+        else:
+            pdata += mhelp
+            if isinstance(val, (int, long)):
+                mtype = "# TYPE {} counter\n".format(key)
+            if isinstance(val, float):
+                mtype = "# TYPE {} gauge\n".format(key)
+            pdata += mtype
+            pdata += "{} {}\n".format(key, val)
+    return pdata
+
+
+def flattenStats(cpdata):
+    "Flatten cherrypy stats to make them suitable for monitoring"
+    if isinstance(cpdata, str):
+        cpdata = json.loads(cpdata)
+    data = {}
+    for key, val in cpdata.iteritems():
+        if key.lower().find('cherrypy') != -1:
+            for kkk, vvv in val.iteritems():
+                nkey = 'cherrypy_app_%s' % kkk
+                nkey = nkey.lower().replace(" ", "_").replace("/", "_")
+                data[nkey] = vvv
+        if key.lower().find('cheroot') != -1:
+            for kkk, vvv in val.iteritems():
+                if kkk == 'Worker Threads':
+                    wdata = []
+                    for wkey, tval in vvv.iteritems():
```

Comment:
Same comment as above regarding the variable names for `wkey` and `tval`. Since here it already becomes clear there is a third level in the CherryPy data structure we are considering, I would, for sure, make those variable names more descriptive.

Reply:
ok, replaced to workerKey and threadValue.